File: Search/whatFlavors.js
/**
 *
 * Author: Dmitriy Shin
 * Date: June 2, 2020
 * Program name: Hash Tables: Ice Cream Parlor
 * 
 */

// TODO Work on the performance
const whatFlavors = (cost = [], money = 0) => {
	const sorted = [ ...cost ].sort((a, b) => a - b)

	const largestIndex = sorted.findIndex((n) => n >= money)
	const sliced = sorted.slice(0, largestIndex)

	const output = []
	for (let i = 0; i < sliced.length; i++) {
		for (let j = i + 1; j < sliced.length; j++) {
			if (sliced[i] + sliced[j] === money) {
				const firstIndex = cost.findIndex((n) => n === sliced[i])
				const secondIndex = cost.findIndex(
					(n, index) => n === sliced[j] && index !== firstIndex
				)

				output.push(firstIndex + 1)
				output.push(secondIndex + 1)
			}
		}
	}

	return output
}

console.log(whatFlavors([ 2, 1, 3, 5, 6 ], 5)) // [1, 3]
console.log(whatFlavors([ 1, 4, 5, 3, 2 ], 4)) // [1, 4]
console.log(whatFlavors([ 2, 2, 4, 3 ], 4)) // [1, 2]

File: Search/whatFlavors.java
import java.util.HashMap;

/**
 *
 * Author: Dmitriy Shin
 * Date: June 8, 2020
 * Program name: Hash Tables: Ice Cream Parlor
 * 
 */

class IceCream {
	public static void whatFlavors(int money, int[] cost) {
		int[] output = new int[2];
		HashMap<Integer, Integer> map = new HashMap<>();
		for (int i = 0; i < cost.length; i++) {
			int x = cost[i];
			int y = money - x;
			Integer j = map.get(y);
			if (j != null) {
				output[0] = i + 1;
				output[1] = j + 1;
				break;
			}
			map.put(x, i);
		}
		System.out.println(String.format("%d %d", output[1], output[0]));
	}
}